File: src/client/java/fr/hardel/asset_editor/client/javafx/components/ui/codeblock/JsonCodeBlockHighlighter.java
package fr.hardel.asset_editor.client.javafx.components.ui.codeblock;

import fr.hardel.asset_editor.client.javafx.lib.highlight.Highlight;
import fr.hardel.asset_editor.client.javafx.lib.highlight.HighlightPalette;
import fr.hardel.asset_editor.client.javafx.lib.highlight.HighlightRegistry;
import fr.hardel.asset_editor.client.javafx.lib.highlight.HighlightStyle;
import fr.hardel.asset_editor.client.javafx.lib.utils.JsonTokenizer;
import javafx.scene.paint.Color;

import java.util.EnumMap;
import java.util.Map;

public final class JsonCodeBlockHighlighter implements CodeBlockHighlighter {

    public static final String STRING = "json-string";
    public static final String NUMBER = "json-number";
    public static final String BOOLEAN = "json-boolean";
    public static final String NULL = "json-null";
    public static final String PROPERTY = "json-property";
    public static final String PUNCTUATION = "json-punctuation";

    @Override
    public void apply(String text, HighlightRegistry registry) {
        Map<JsonTokenizer.TokenType, Highlight> highlightsByType = new EnumMap<>(JsonTokenizer.TokenType.class);

        for (JsonTokenizer.Token token : JsonTokenizer.tokenize(text)) {
            if (token.type() == JsonTokenizer.TokenType.WHITESPACE)
                continue;

            Highlight highlight = highlightsByType.computeIfAbsent(token.type(), ignored -> new Highlight());
            highlight.add(token.start(), token.end());
        }

        for (Map.Entry<JsonTokenizer.TokenType, Highlight> entry : highlightsByType.entrySet()) {
            String name = entry.getKey().highlightName();
            if (name != null)
                registry.set(name, entry.getValue());
        }
    }

    public static void installDefaultPalette(HighlightPalette palette) {
        palette.set(STRING, HighlightStyle.foreground(Color.web("#98c379")));
        palette.set(NUMBER, HighlightStyle.foreground(Color.web("#d19a66")));
        palette.set(BOOLEAN, HighlightStyle.foreground(Color.web("#56b6c2")));
        palette.set(NULL, HighlightStyle.foreground(Color.web("#c678dd")));
        palette.set(PROPERTY, HighlightStyle.foreground(Color.web("#61afef")));
        palette.set(PUNCTUATION, HighlightStyle.foreground(Color.web("#abb2bf")));
    }
}

File: src/client/kotlin/fr/hardel/asset_editor/client/compose/components/ui/codeblock/JsonCodeBlockHighlighter.kt
package fr.hardel.asset_editor.client.compose.components.ui.codeblock

import fr.hardel.asset_editor.client.compose.StudioColors
import fr.hardel.asset_editor.client.compose.lib.highlight.Highlight
import fr.hardel.asset_editor.client.compose.lib.highlight.HighlightPalette
import fr.hardel.asset_editor.client.compose.lib.highlight.HighlightRegistry
import fr.hardel.asset_editor.client.compose.lib.highlight.HighlightStyle
import fr.hardel.asset_editor.client.compose.lib.utils.JsonTokenizer

class JsonCodeBlockHighlighter : CodeBlockHighlighter {

    override fun apply(text: String, registry: HighlightRegistry) {
        val highlightsByType = LinkedHashMap<JsonTokenizer.TokenType, Highlight>()

        for (token in JsonTokenizer.tokenize(text)) {
            if (token.type == JsonTokenizer.TokenType.WHITESPACE) {
                continue
            }

            val highlight = highlightsByType.getOrPut(token.type) { Highlight() }
            highlight.add(token.start, token.end)
        }

        for ((type, highlight) in highlightsByType) {
            val name = type.highlightName ?: continue
            registry.set(name, highlight)
        }
    }

    companion object {
        const val STRING = "json-string"
        const val NUMBER = "json-number"
        const val BOOLEAN = "json-boolean"
        const val NULL = "json-null"
        const val PROPERTY = "json-property"
        const val PUNCTUATION = "json-punctuation"

        fun installDefaultPalette(palette: HighlightPalette) {
            palette.set(STRING, HighlightStyle.foreground(StudioColors.SyntaxString))
            palette.set(NUMBER, HighlightStyle.foreground(StudioColors.SyntaxNumber))
            palette.set(BOOLEAN, HighlightStyle.foreground(StudioColors.SyntaxBoolean))
            palette.set(NULL, HighlightStyle.foreground(StudioColors.SyntaxNull))
            palette.set(PROPERTY, HighlightStyle.foreground(StudioColors.SyntaxProperty))
            palette.set(PUNCTUATION, HighlightStyle.foreground(StudioColors.SyntaxPunctuation))
        }
    }
}
